### ISSUE TITLE
Conch's public key authentication process is confusing.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3537]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#3537|
|Type|enhancement|
|Created|2008-11-15 22:35:53Z|
|Branch|https://github.com/twisted/twisted/tree/valid-public-key-3537|

Specifically, the public key checker raises `ValidPublicKey` to indicate that it received a request to *try* public key authentication. The userauth code expects that exception to be raised, and sends an "public key is OK" message before ending that round of authentication (the client is expected to attempt another).

As page 9 of http://www.ietf.org/rfc/rfc4252.txt shows, the way the client indicates that it wants to try public key auth is by sending a regular public key auth packet request without the signature.

I think the right thing to do here is change:

```python
class ValidPublicKey(Exception):
    """This is thrown during the authentication process if the public key
    is valid for the user.
    """
```

to 

```python
class ValidPublicKey(Exception):
    """
    Raised by public key checkers when they receive public key credentials that don't
    contain a signature at all, but are valid in every other way. (e.g. the public key
    matches one in the user's authorized_keys file). 

    L{SSHUserAuthServer} will catch this exception and tell the user to re-authenticate,
    this time with a signature.

    See http://www.ietf.org/rfc/rfc4252.txt section 7 for more details.
    """
```

<details><summary>Searchable metadata</summary>

```
trac-id__3537 3537
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_valid_public_key_3537 branches/valid-public-key-3537
branch_author__jml__exarkun jml, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__documentation documentation
time__1226788553000000 1226788553000000
changetime__1227105942000000 1227105942000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @jml|
|-|-|

Raising `ValidPublicKey` from the `requestAvatarId` method of an `ICredentialsChecker` implementation violates the interface.  `UnauthorizedLogin` is the exception documented as being allowed to come from `requestAvatarId`.

Making `ValidPublicKey` a subclass of `UnauthorizedLogin` would fix this.  Then, in addition to improving the documentation of `ValidPublicKey`, I would recommend also improving the documentation of `SSHPublicKeyDatabase` to document the circumstance under which it raises `ValidPublicKey`.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [25359]) Branching to 'valid-public-key-3537'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @jml|
|-|-|

Thanks a lot for the clarifications. I have some comments:

 1.
```
+    L{SSHUserAuthServer} will catch this exception and tell the user to
```
Using `L{}` on something not imported in the module doesn't generate a link to the object. You have to use something like that:  `L{SSHUserAuthServer<twisted.conch.ssh.userauth.SSHUserAuthServer>}`
 2.
```
+    See http://www.ietf.org/rfc/rfc4252.txt section 7 for more details.
```
Please use `U{}` to make a clickable link.
 3.
```
+        the credentials do not include a signature. See L{ValidPublicKey} for
```
Same as first point. Importing `ValidPublicKey` instead of error could be nice, though.
 4. `_ebRequestAvatarId` doesn't need to check for both `UnauthorizedLogin` and `ValidPublicKey` anymore. Actually, it looks like this method may be completely useless now?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25391]) address review comments 1 and 2

refs [#3537](https://github.com/twisted/twisted/issues/3537)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [25392]) address review comment 3

refs [#3537](https://github.com/twisted/twisted/issues/3537)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25393]) improve the docs for ISSHPrivateKey

refs [#3537](https://github.com/twisted/twisted/issues/3537)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25394]) balance parens

refs [#3537](https://github.com/twisted/twisted/issues/3537)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [25395]) don't check for ValidPublicKey, just check for UnauthorizedLogin

refs [#3537](https://github.com/twisted/twisted/issues/3537)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Fixed all points, I think.  `_ebRequestAvatarId` is still there, though, since it does stuff with logging and transforming the type of the failure.  It's kind of a crummy method, and it has poor test coverage (nothing exercises the inside of the if suite, but I suspect it can run in real life).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @exarkun|
|-|-|

OK, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [25439]) Merge valid-public-key-3537

Author: jml, exarkun
Reviewer: exarkun, therve
Fixes: [#3537](https://github.com/twisted/twisted/issues/3537)

Clarify the API documentation relating to `ValidPublicKey` in Conch's SSH
key checker so that it explains how this failure from a login attempt indicates
that another login attempt with the same key and a signature may actually
succeed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
